tests/tokens/test_resource_access.py

```diff
@@ -0,0 +1,699 @@
+
+from invenio_rdm_records.records import RDMParent, RDMRecord
+from invenio_rdm_records.tokens import ResourceAccessToken
+from invenio_rdm_records.tokens.errors import (
+    ExpiredTokenError,
+    InvalidTokenError,
+    InvalidTokenIDError,
+    MissingTokenIDError,
+)
+from invenio_rdm_records.tokens.scopes import tokens_generate_scope
+
+
+def _generate_pat_token(
+    db,
+    uploader,
+    client,
+    access_token,
+    scope=tokens_generate_scope.id,
+    expires=datetime.utcnow() + timedelta(hours=10),
+):
+    """Create a personal access token."""
+    with db.session.begin_nested():
+        token_ = Token(
+            client_id=client,
+            user_id=uploader.id,
+            access_token=access_token,
+            expires=expires,
+            is_personal=False,
+            is_internal=True,
+            _scopes=scope,
+        )
+        db.session.add(token_)
+    db.session.commit()
+    return dict(
+        token=token_,
+        auth_header=[
+            ("Authorization", "Bearer {0}".format(token_.access_token)),
+        ],
+    )
+
+
+def _rat_gen(token, payload=None, headers=None):
+    """Create a resource access token."""
+    payload = {"iat": datetime.utcnow()} if payload is None else payload
+    headers = {"kid": str(token.id)} if headers is None else headers
+    return jwt.encode(
+        payload=payload,
+        key=token.access_token,
+        algorithm="HS256",
+        headers=headers,
+    )
+
+
+def create_record_w_file(client, record, headers, is_published=True):
+    """Create record with a file."""
+    # Create draft
+    record["files"] = {"enabled": True}
+    response = client.post("/records", json=record, headers=headers)
+    assert response.status_code == 201
+    recid = response.json["id"]
+
+    # Attach a file to it
+    response = client.post(
+        f"/records/{recid}/draft/files", headers=headers, json=[{"key": "test.pdf"}]
+    )
+    assert response.status_code == 201
+    response = client.put(
+        f"/records/{recid}/draft/files/test.pdf/content",
+        headers={
+            "content-type": "application/octet-stream",
+            "accept": "application/json",
+        },
+        data=BytesIO(b"testfile"),
+    )
+    assert response.status_code == 200
+    response = client.post(
+        f"/records/{recid}/draft/files/test.pdf/commit", headers=headers
+    )
+    assert response.status_code == 200
+
+    # Publish it
+    if is_published:
+        response = client.post(
+            f"/records/{recid}/draft/actions/publish", headers=headers
+        )
+        assert response.status_code == 202
+
+    return recid
+
+
+def test_rat_validation_failed(
+    app, db, users, uploader, superuser_identity, minimal_record, oauth2_client
+):
+    """Test possible validation failings of Resource access tokens."""
+    parent = RDMParent.create({"access": {"owned_by": [{"user": users[0].id}]}})
+    record = RDMRecord.create(minimal_record, parent=parent)
+
```

Comment:
nit: add a comment why each case should fail, to ease reading (e.g. file does not exist, not owner, wrong kid, etc.). In addition, there are many cases in this one, when if one fails the subsequent wont' be tested. Might be worth spliting to one test per check.